# SCIKIT LEARN WORKFLOW
## END TO END SCIKIT LEARN WORKFLOW

What is end to end means? 

One thing to note though Scikit Learn is such big library thus only call what you need will be a better approach to practical use.

### Random Forest Classifier Workflow for Classifying Heart Disease.
#### 1. Get the data ready
I will use the heart disease data. Remember that it was stored in csv format. First I need to import it first. Pandas will be the appropriate library to get and process the data to form Data Frame.

In [2]:
# get the data heart-disease.csv
import pandas as pd
heart_disease_data = pd.read_csv('../data/heart-disease.csv')
heart_disease_data

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  \
0     63    1   3       145   233    1        0      150      0      2.3   
1     37    1   2       130   250    0        1      187      0      3.5   
2     41    0   1       130   204    0        0      172      0      1.4   
3     56    1   1       120   236    0        1      178      0      0.8   
4     57    0   0       120   354    0        1      163      1      0.6   
..   ...  ...  ..       ...   ...  ...      ...      ...    ...      ...   
298   57    0   0       140   241    0        1      123      1      0.2   
299   45    1   3       110   264    0        1      132      0      1.2   
300   68    1   0       144   193    1        1      141      0      3.4   
301   57    1   0       130   131    0        1      115      1      1.2   
302   57    0   1       130   236    0        0      174      0      0.0   

     slope  ca  thal  target  
0        0   0     1       1  
1        0   0     2       1  
2        2   0     2       1  
3        2   0     2       1  
4        2   0     2       1  
..     ...  ..   ...     ...  
298      1   0     3       0  
299      1   0     3       0  
300      1   2     3       0  
301      1   1     3       0  
302      1   1     2       0  

[303 rows x 14 columns]

***NOTE***
The heart_disease_data consist of all patient characteristics except for the target column. Target column is the end result whether the patient has heart_disease or not and we want to seek the connection with each character.

In [24]:
# first we need to separate the target column from other data
x_data = heart_disease_data.drop('target', axis=1)
x_data.head()

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  \
0   63    1   3       145   233    1        0      150      0      2.3      0   
1   37    1   2       130   250    0        1      187      0      3.5      0   
2   41    0   1       130   204    0        0      172      0      1.4      2   
3   56    1   1       120   236    0        1      178      0      0.8      2   
4   57    0   0       120   354    0        1      163      1      0.6      2   

   ca  thal  
0   0     1  
1   0     2  
2   0     2  
3   0     2  
4   0     2

In [25]:
# now the target label will be in the y_data
y_data = heart_disease_data['target']
y_data.head()

0    1
1    1
2    1
3    1
4    1
Name: target, dtype: int64

***NOTE:***
now I need to split the data frame in the x_data and y_data to do this I will use the scikit learn libs

The way to import is a bit different since it is named sklearn thus to import I need import sklearn.

Now as this library is quite bid I need to be specific. This time it is model_selection to be exact I want to randomly split the data frame heart_disease_data y_data and x_data. I will use the train_test_split library

In [26]:
# import the library
# https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html
from sklearn.model_selection import train_test_split

# just to be safe if we need to handle some matrices I will import numpy as well
import numpy as np

In [27]:
x_train, x_test, y_train, y_test = train_test_split(x_data, y_data)
# let's test the result but since too big I will just validate each size
x_train.shape, y_train.shape, x_test.shape, y_test.shape

((227, 13), (227,), (76, 13), (76,))

***NOTE:***
The data (shape) above is important. 

I need to validatae that the x_train and y_train first size is the same (in this case 227).

The same case is true for the x_test and y_test (in this case 76)

It means will be 227 data for train and 76 for testing that chosen randomly.

#### 2. Choose the model and hyperparameters
Now this is the very heart of the Scikit Learn. A variety of models or in machine learning is known as classifiers. However, in Scikit documentation it is more commonly known as estimators. 

TODO So what is hyperparameters??

Next we will use Random Forest Classifier: https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestClassifier.html?highlight=forest#sklearn.ensemble.RandomForestClassifier

TODO learn about Random Forest Classifier model.